### DISCUSSION TITLE
Asyncio: Polling + ISR's...  Bad combination

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#11427] [General]
Hello everybody!

I've been reading that is not recommended to use ISR interruptions in general when you are using asyncio, for multiple important reasons... (some of them I understand, others not. I'm a junior junior programmer, sorry). but I thought that it would be very usefull in some simple cases with a just little bit of code in specific cases. For example:

Executing this code (synchronous, without asyncio):

    from machine import Pin # type:ignore
    from time import sleep

    motion = False

    def handle_interrupt(pin):
        global motion
        motion = True
        global interrupt_pin
        interrupt_pin = pin 

    led = Pin(2, Pin.OUT)
    pir = Pin(36, Pin.IN)

    pir.irq(trigger=Pin.IRQ_RISING, handler=handle_interrupt)

    while True:
        if motion:
        print('Motion detected! Interrupt caused by:', interrupt_pin)
        led.value(1)
        sleep(20)
        led.value(0)
        print('Motion stopped!')
        motion = False

is completly fine and it works.

But using the same strategy inside a task/coro doesn't work:

    async def Device():
    ...
        a_boolean = False

        def ISR(Pin):
            global a_boolean
            a_boolean = True

        in2.irq(handler = ISR , trigger = machine.Pin.IRQ_RISING)

        while True:
            if a_boolean:
                print("Very well")

The scope of "a boolean" variable is never global, so "a boolean" always is False and we don't come into "if". What I'm doing wrong?? Maybe this is one important reason for not use this mix: asyncio + isr? Is there a solution?

Thank you very much.

**Comment by dlech** [comment]
https://github.com/peterhinch/micropython-async/blob/master/v3/docs/INTERRUPTS.md

**Comment by bixb922** [comment]
This has no relationship with asyncio. The ```a_boolean = False``` declares a local variable in the scope of Device() which is different from ```global a_boolean```. If in doubt, please revisit python scoping rules.

Don't use global variables to communicate events between ISR and main loop. The ThreadSafeFlag has been specifically designed to do that signaling safely as described in the links provided above by @dlech. Please see the examples there. 

Just curiosity: What is the reaction time you need for the interrupt?

**Comment by bixb922** [comment]
Just to clarify: "Polling a device" means a code like this:
```py
    while True:
        while pin.value()==0:
            await asyncio.sleep_ms( 100 )
        ... Do something because the signal went from 0 to 1
        while pin.value()==1:
            await asyncio.sleep_ms( 100 )
```
No ISR is needed when polling.

**Reply by peterhinch** [reply]
Or use the [Switch class](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md) which gives faster response to a state change while debouncing.

Using interrupts to handle a mechanical switch is rarely good practice.

**Reply by nastiliano** [reply]
Hello @bixb922 , @peterhinch . A pleasure.

Mmmm... I think I was completly wrong with my code...
I was basing polling with the following strategy:

    async def Device():
        #Local variables (I don't know If this is correct. I mean, if it's correct to put variables here in order to avoid the "creation"
        # of local variables each time a coro is called. I put a "print()" and it's only executed once... so I put my NVM reading
        #variables here...)
        ...

    while True:
        await uasyncio.sleep_ms(50)

        async def anti_deboucing:
            while pin1.value() == 0:
                await uasyncio.sleep_ms(10)

        if pin1.value() == 0:
            try:
                await uasyncio.wait_for(anti_deboucing() , timeout = 1)
            except:
                pass

            #Starting temporized relay
            while relay1_time:

                #Stop the relay immediately (A safety INPUT. For this reason I wanted to use ISR. High priority. Stop all execution for
                #security)
                if "a_booblean": #A boolean catched from ISR handler
                    #reset variables
                    ...
                    break

                await uasyncio.sleep_ms(100)
                realy1_time -= 1
                reley1.value(0)

               #If countingdown equals 0 
               if relay1_time == 0:
                   relay.value(1)
                   #reset variables
                   ...
                   break

Could you help me to rewritting just a little bit this code using REAL polling and "official" Switch class???

This is what I'm looking for. External switches (without debouncig but obviouslu without blocking my program execution. just like a pulse, a falling flanck) that if pin1 is pressed a relay is working with a countingdown, but if a pin2 is closed during the execution the relay stops, even more, if pin2 is closed the temporized relay never has to run. (for this reason I was planning my code in that way... The pin2 is special, is not just a switch, is a safety input).

Thank you, thank you very much.
Best regards.

**Reply by peterhinch** [reply]
The word "safety" is a red light. If you are talking about human safety (from injury) different rules apply. Please clarify what you are doing before we go any further.

**Reply by nastiliano** [reply]
Hello, @peterhinch 

When I say safety, yes, I say human safety.
I'm programing an opening/closing automated gate. Is a "stop" safety operation, but I think that @bixb922 said to me that if I'm no need something minor to 50ms, polling is a good choice. I want to stop the gate maneuver immediatly, but I think "ms" is enought, I think I don't need something like "us" in my case.

Thanks again.

**Reply by bixb922** [reply]
If the solution might affect human safety in any way, or tries to prevent injury, I think a specialized design must be made. I really do not know about such solutions, but do I know they need very careful designs. Whatever I have expressed before does not apply to solutions that might affect human safety, so please do not take them into account for a design of this type.

**Reply by peterhinch** [reply]
I concur. I cannot offer any support for safety-critical design as I'm not qualified to do so. @nastiliano You need specialist advice.

**Reply by nastiliano** [reply]
Uhmmmm...ok, ok.

I try to program a simple movement of an automated door. I thought it was not so dangerous or specialized thing. Of course, I will do all of the things "the specialists" say to me, because all security is little to avoid injuries. But the movement of the gate is very slow and (more important), my device will have 2 securities: an input for stop immediately the gate (based on fotocells) and a "current sensor" to prevent somebody could be stuck (in a overload, stops the gate and go back).
Anyway, would there be some "specialist" in these forums?

Thank you very very much.
Best regards.

**Reply by 2dof** [reply]
If You have possibility try read  some iso standards  i.e  **EN 1155** (BS EN 1155:2003-04 ) (for electromechanical or electromagnetic stand-alone doors) 

Unfortunately iso docs are quite expensive, but if your work is just a programmer implementing requirements based on project documentation if there is no any, try contact with specialists responsible for project (electro-mechanical part), they should have access to ISO docs.

 edit: in ww you fill find some information  like [ example test report](https://www.safelincs.co.uk/files/Salamander_EN1155_-_151781.pdf) , where You can find some info about safety

**Reply by jimmo** [reply]
> Anyway, would there be some "specialist" in these forums?

No, and this is not the place for getting free professional advice on the implementation of safety-critical systems. You need to discuss this with whomever has assigned you this task.

**Reply by nastiliano** [reply]
> > Anyway, would there be some "specialist" in these forums?
> 
> No, and this is not the place for getting free professional advice on the implementation of safety-critical systems. You need to discuss this with whomever has assigned you this task.

Are you ok?

1st: I'm not asking for nothing free or commercial to anybody. I'm asking for HELP

2nd: As a maintainer, look to be a little bit polite. Thanks  (first, hello...  thank you. This kind of things)

**Reply by peterhinch** [reply]
There are serious legal issues here. @bixb922, @jimmo and myself are aware of these. Are you?

**Reply by nastiliano** [reply]
> There are serious legal issues here. @bixb922, @jimmo and myself are aware of these. Are you?

I completly understood before, for this reason I stopped to reply more in this post and I'm looking for other ways for help and I'm trying to learn by myself.
And I'm completly thankful to you (@bixb922, @jimmo, @peterhinch), but I'm annoyed with someone to say me that (trying to find " I don't know what" for free... that's all).
Thanks again to you and all communty.
Best regards.

**Reply by 2dof** [reply]
@ nastiliano  You should not be offended by @jimmo or other users.  I'm sure that " not the place for free professional advice on the implementation of safety-critical systems" means that  professional/specialists will be always carefull to give pure technical ( mark by his name) advices in public forum (even if there will be professional advice ) - because this is not the place for that (it is not professional).    

Just like @peterhinch said  there legal issues (in safety-critical systems ) and always should be discuss in Your Team/organisation and this is 
very good advice  given by @jimmo 

if You are just a programmer (without experience in automation in safety-critical systems)  just demand from klient/ menager  a "entry" range abut designed and desired closing speed etc ( it should be already done in hardware project), and as programmer just implement your system to be flexible for changes (adjusting etc). 

-  if You do "all of the things "the specialists" say to me"  do always on "paper" (signed by "the specialists").  
- if knowing a ISO/standard docs are out of your scope scope of work (simply you do not need know it  as programmer) it is always good to read ones it always helps.

**Reply by nastiliano** [reply]
> @ nastiliano You should not be offended by @jimmo or other users. I'm sure that " not the place for free professional advice on the implementation of safety-critical systems" means that professional/specialists will be always carefull to give pure technical ( mark by his name) advices in public forum (even if there will be professional advice ) - because this is not the place for that (it is not professional).
> 
> Just like @peterhinch said there legal issues (in safety-critical systems ) and always should be discuss in Your Team/organisation and this is very good advice given by @jimmo
> 
> if You are just a programmer (without experience in automation in safety-critical systems) just demand from klient/ menager a "entry" range abut designed and desired closing speed etc ( it should be already done in hardware project), and as programmer just implement your system to be flexible for changes (adjusting etc).
> 
> * if You do "all of the things "the specialists" say to me"  do always on "paper" (signed by "the specialists").
> * if knowing a ISO/standard docs are out of your scope scope of work (simply you do not need know it  as programmer) it is always good to read ones it always helps.

Hello @2dof,

Thanks for your response.

I'm not in any organization or similar. I'm an enthusiast/hobbiest that I want to implement a gate automation for ME, not for earn money and I don't need so much specification to implement my hobby, it's for me.

I only was asking for ISR's a little bit more, because my problem is that using "polling" for me is complicated if I want an implementation of "Stop switch" that stops all execution. I wanted to use polling but I wanted an interruption to stop all execution immediately, that's all.

Thanks for all your answers.

Best regards.

**Reply by 2dof** [reply]
So if You want to just "stop switch" , go to  link given by @peterhinch  ( [Switch class](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md#31-switch-class) in his tutorial (also read other uasyncio related issues [Linking uasyncio and other contexts](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md#12-soft-interrupt-service-routines)

Aware that  ""Stop switch" that stops all execution"  means  a " kill switch" ( == power off -> another protection  mechanism), but That what are you doing is controlling motor drive and all other (protection) sequences should be running. 

I do not know what type of gate You want control (size, weigh etc), but that what You implementing is control unit for drive gate motor with protection it should be done with time sampling mechanism if You want to control gate speed too, then all sensors/protection switches have to be check in the control loop.

**Reply by rolandvs** [reply]
The STM timers/pwm have a feature to kill the pwm using a single gpio, independent from software. It can be configured in mpy. This is a nice feature to have.

**Reply by nastiliano** [reply]
> So if You want to just "stop switch" , go to link given by @peterhinch ( [Switch class](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md#31-switch-class) in his tutorial (also read other uasyncio related issues [Linking uasyncio and other contexts](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md#12-soft-interrupt-service-routines)
> 
> Aware that ""Stop switch" that stops all execution" means a " kill switch" ( == power off -> another protection mechanism), but That what are you doing is controlling motor drive and all other (protection) sequences should be running.
> 
> I do not know what type of gate You want control (size, weigh etc), but that what You implementing is control unit for drive gate motor with protection it should be done with time sampling mechanism if You want to control gate speed too, then all sensors/protection switches have to be check in the control loop.

Hello again.

Exactly, this is what I want.
For example, my code was something like (I'm inventing and improvising):

    ...
    def async Device():
        in_1= machine.Pin(36 , machine.Pin.IN)
        
        def stop(pin):
            print("hello?")
            global boolean_variable
            boolean_variable = True
        stop = machine.Pin(36 , machine.Pin.IN)
        stop.irq(handler = stop , trigger = machine.Pin.IRQ_FALLING)

        while True:
            if boolean_variable:
                boolean_variable = False
                break

             if in1.value() == 0:
                 if boolean_variable:
                     boolean_variable = False
                     break
                #Start OPEN relay...

        await uasyncio.sleep_ms(50)

But this simple code doesn't work:
    -1st: Interrupt is handled forever, when I execute this code, prints: hello, hello, hello, hello...
    -2nd: A variable scope error occurs. It seems like "global" parameter don't work

I'm stuck here for weeks :)

**Reply by 2dof** [reply]
in Your code:
- 2 times you defining ``` machine.Pin(36 , machine.Pin.IN) ``` as input   ```in_1```  and   ``` stop ```   as this same pin  and  you are using 
  these two objects: as irq source  and for reading pin value  - not need do that. 
- do not define inside  async function  pins , better outside or in class object which will be your devise , all other functions ( irq handler, async can be then a define as members of class 
-  ```global boolean_variable```  is declared in  stop(pin) function but should be also in  ```def async Device()``` 
- do not use ``` break``` on if condition - make no sense since it is already end of  'if' 

If source of interrupt is a mechanical (switch)  it will be multiple triggering iqr  because of connection vibrations , better learn how to use 
 [Switch class](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md#31-switch-class) -  as example it is almost ready solution for your code ( you can define close_func()  as your stop() , and open_func()  to reset boolean_variable 

 see my demo (based on example from switch doc):
``` 
from machine import Pin  
import time
from primitives import Switch
import uasyncio as asyncio

stop_flag = False

async def closed_switch():
    global stop_flag
    stop_flag = True 
    print('switch closed',stop_flag )

async def opened_switch():
    global stop_flag
    stop_flag = False 
    print('switch opened',stop_flag)
    
async def my_app():
    global stop_flag 
    stop_flag = False
    pin = Pin(4, Pin.IN, Pin.PULL_UP)                    # 
    sw = Switch(pin)
    sw.close_func(closed_switch)   
    sw.open_func(opened_switch)    
    await asyncio.sleep(100)  # Dummy application code

print('run')
asyncio.run(my_app())  # Run main application code
```
 I said earlier that better define gpio outside function, but in this case ```my_app()``` is my main application (not a 'device')  
 
Note that  Switch class and other depended primitives form uasyncio drivers library must be downloaded and loaded to your 
device ( they are not installed in uasyncio).

**Reply by nastiliano** [reply]
This is what I wanted! Thank you.

I re-adapted your code to mine and is working good with "polling"!
Now I can poll the device and make "important" interruptions.

I'm going to follow with my work :)

Thanks again to everybody.

Best regards.
